src/Data/Equality/Utils/SizedList.hs

```diff
@@ -0,0 +1,64 @@
+{-# LANGUAGE DeriveTraversable #-}
+{-# LANGUAGE TypeFamilies #-}
+{-|
+   Util: A list with a O(1) size function
+ -}
+module Data.Equality.Utils.SizedList where
+
+import GHC.Exts
+import Data.Foldable
+
+-- | A list with O(1) size access and O(1) conversion to normal list
+data SList a = SList ![a] {-# UNPACK #-} !Int
+  deriving Traversable
+
+instance Semigroup (SList a) where
+  (<>) (SList a i) (SList b j) = SList (a <> b) (i+j)
+  {-# INLINE (<>) #-}
+
+instance Monoid (SList a) where
+  mempty = SList mempty 0
+  {-# INLINE mempty #-}
+
+instance Functor SList where
+  fmap f (SList a i) = SList (fmap f a) i
+
+instance Foldable SList where
+  fold       ( SList l _) = fold l
+  foldMap f  ( SList l _) = foldMap f l
+  foldMap' f ( SList l _) = foldMap' f l
+  foldr f b  ( SList l _) = foldr f b l
+  foldr' f b ( SList l _) = foldr' f b l
+  foldl f b  ( SList l _) = foldl f b l
+  foldl' f b ( SList l _) = foldl' f b l
+  foldr1 f   ( SList l _) = foldr1 f l
+  foldl1 f   ( SList l _) = foldl1 f l
+  toList     ( SList l _) = l
+  null       ( SList l _) = null l
```

Comment:
Make sure to use the `null` from `Data.List`. It's going to be a tiny bit faster.